Error when executing liberty:run: Cannot find 'initializeInOrder'

I get an error when using looseApplication set to false during the call to execute mojo maven-ear-plugin

```[ERROR] Failed to execute goal io.openliberty.tools:liberty-maven-plugin:3.3.5-M2-SNAPSHOT:run (default-cli) on project AIRPISYNCEnterprise: Unable to execute mojo: Unable to parse configuration of mojo org.apache.maven.plugins:maven-ear-plugin:3.0.2:ear for parameter initializeInOrder: Cannot find 'initializeInOrder' in class org.apache.maven.plugins.ear.EarMojo -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.openliberty.tools:liberty-maven-plugin:3.3.5-M2-SNAPSHOT:run (default-cli) on project AIRPISYNCEnterprise: Unable to execute mojo
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Unable to execute mojo
	at org.codehaus.mojo.pluginsupport.MojoSupport.execute(MojoSupport.java:129)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	... 20 more
Caused by: org.apache.maven.plugin.MojoExecutionException: Unable to execute mojo
	at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo(MojoExecutor.java:122)
	at io.openliberty.tools.maven.server.StartDebugMojoSupport.runMojo(StartDebugMojoSupport.java:190)
	at io.openliberty.tools.maven.server.RunServerMojo.doExecute(RunServerMojo.java:58)
	at org.codehaus.mojo.pluginsupport.MojoSupport.execute(MojoSupport.java:122)
	... 22 more
Caused by: org.apache.maven.plugin.PluginConfigurationException: Unable to parse configuration of mojo org.apache.maven.plugins:maven-ear-plugin:3.0.2:ear for parameter initializeInOrder: Cannot find 'initializeInOrder' in class org.apache.maven.plugins.ear.EarMojo
	at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:665)
	at org.apache.maven.plugin.internal.DefaultMavenPluginManager.getConfiguredMojo(DefaultMavenPluginManager.java:597)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:124)
	at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo(MojoExecutor.java:120)
	... 25 more
Caused by: org.codehaus.plexus.component.configurator.ComponentConfigurationException: Cannot find 'initializeInOrder' in class org.apache.maven.plugins.ear.EarMojo
	at org.eclipse.sisu.plexus.CompositeBeanHelper.setProperty(CompositeBeanHelper.java:252)
	at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.processConfiguration(ObjectWithFieldsConverter.java:101)
	at org.codehaus.plexus.component.configurator.BasicComponentConfigurator.configureComponent(BasicComponentConfigurator.java:34)
	at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:635)
	... 28 more```
If you look at the passed configuration (see line 747 in attached maven.log file) you see clearly that the initializeInOrder is indeed in the passed configuration but should not be required.  What confuses me is that parameter is not used in the ear goal but rather in the generate-application-xml goal.

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR/pom.xml
@@ -61,6 +61,7 @@
                         automatically have any modules (EJB-JARs and WARs) use them -->
                     <defaultLibBundleDir>lib</defaultLibBundleDir>
                     <outputFileNameMapping>@{artifactId}@.@{extension}@</outputFileNameMapping>
+                    <initializeInOrder>true</initializeInOrder>
                     <modules>
                         <webModule>
                             <groupId>io.openliberty.tools.it</groupId>
@@ -89,7 +90,7 @@
                     </assemblyArtifact>
                     <serverName>test</serverName>
                     <appsDirectory>apps</appsDirectory>
-                    <looseApplication>true</looseApplication>
+                    <looseApplication>false</looseApplication>
                     <include>usr</include>
                 </configuration>
                 <executions>

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/pom.xml
@@ -9,7 +9,7 @@
         <version>1.0-SNAPSHOT</version>
     </parent>
 
-    <artifactId>SampleEAR</artifactId>
+    <artifactId>SampleEAR_NonLooseApp</artifactId>
 
     <packaging>ear</packaging>
 

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/pom.xml
@@ -13,7 +13,7 @@
 
     <packaging>ear</packaging>
 
-    <name>ear-project-it:  SampleEAR Module</name>
+    <name>ear-project-it:  SampleEAR_NonLooseApp Module</name>
 
     <dependencies>
         <dependency>
@@ -90,7 +90,7 @@
                     </assemblyArtifact>
                     <serverName>test</serverName>
                     <appsDirectory>apps</appsDirectory>
-                    <looseApplication>true</looseApplication>
+                    <looseApplication>false</looseApplication>
                     <include>usr</include>
                 </configuration>
                 <executions>

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/application/META-INF/license
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/application/META-INF/license
@@ -1,0 +1,15 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2017.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/application/META-INF/license
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/application/META-INF/license
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * (c) Copyright IBM Corporation 2017.
+ * (c) Copyright IBM Corporation 2021.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/liberty/config/server.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEar_NonLooseApp/src/main/liberty/config/server.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<server description="jee7sample test server">
+
+    <!-- Enable features -->
+    <featureManager>
+        <feature>javaee-7.0</feature>
+    </featureManager>
+        
+    <keyStore password="keyPassword"/>
+    <basicRegistry id="basic" realm="BasicRealm">
+        <user name="admin" password="password" />
+    </basicRegistry>
+
+    <!-- To access this server from a remote client add a host attribute to the following element, e.g. host="*" -->
+    <httpEndpoint id="defaultHttpEndpoint"
+                  httpPort="9080"
+                  httpsPort="9443" />
+                  
+</server>

--- a/liberty-maven-plugin/src/it/ear-project-it/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/pom.xml
@@ -17,6 +17,7 @@
         <module>SampleWAR</module>
         <module>SampleWAR2</module>
         <module>AppXmlEAR</module>
+        <module>SampleEar_NonLooseApp</module>
         <module>SampleEAR</module>
         <module>SampleWLP</module>
         <module>skinnywar-ear</module>

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/ExecuteMojoUtil.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/ExecuteMojoUtil.java
@@ -138,6 +138,27 @@ public class ExecuteMojoUtil {
     // https://maven.apache.org/surefire/maven-surefire-report-plugin/failsafe-report-only-mojo.html
     private static final ArrayList<String> FAILSAFE_REPORT_ONLY_PARAMS = REPORT_ONLY_PARAMS;
 
+    // https://maven.apache.org/plugins/maven-ear-plugin/ear-mojo.html
+    private static final ArrayList<String> EAR_PARAMS = new ArrayList<>(
+            Arrays.asList("earSourceDirectory", "outputDirectory", "outputFileNameMapping", "tempFolder", "workDirectory",
+                    "applicationXml", "archive", "artifactTypeMappings", "classifier", "defaultLibBundleDir", "earSourceExcludes",
+                    "earSourceIncludes", "encoding", "escapeString", "escapedBackslashesInFilePath",
+                    "fileNameMapping", "filtering", "filters", "generatedDescriptorLocation",
+                    "includeLibInApplicationXml", "jboss", "mainArtifactId", "modules", "nonFilteredFileExtensions", "outputTimestamp",
+                    "packagingExcludes", "packagingIncludes", "skinnyModules", "skinnyWars", "skipClassPathModification", "unpackTypes",
+                    "useBaseVersion", "useJvmChmod", "version"));
+            
+    // https://maven.apache.org/plugins/maven-war-plugin/war-mojo.html
+    private static final ArrayList<String> WAR_PARAMS = new ArrayList<>(
+            Arrays.asList("outputDirectory", "warSourceDirectory", "webappDirectory", "workDirectory", "archive",
+                    "archiveClasses", "attachClasses", "classesClassifier", "classifier", "containerConfigXML", "delimiters",
+                    "dependentWarExcludes", "dependentWarIncludes", "escapeString", "escapedBackslashesInFilePath",
+                    "failOnMissingWebXml", "filteringDeploymentDescriptors", "filters", "includeEmptyDirectories",
+                    "nonFilteredFileExtensions", "outdatedCheckPath", "outputFileNameMapping", "outputTimestamp", "overlays", "packagingExcludes",
+                    "packagingIncludes", "primaryArtifact", "recompressZippedFiles", "resourceEncoding", "skip", "supportMultiLineFiltering",
+                    "useDefaultDelimiters", "useJvmChmod", "warSourceExcludes", "warSourceIncludes", "webResources", "webXml"));
+
+
     // https://maven.apache.org/plugins/maven-ear-plugin/generate-application-xml-mojo.html
     private static final ArrayList<String> EAR_GENERATE_APPLICATION_XML_PARAMS = new ArrayList<>(
             Arrays.asList("outputFileNameMapping", "tempFolder", "workDirectory", "applicationId", "applicationName",
@@ -250,6 +271,7 @@ public class ExecuteMojoUtil {
     private static Xpp3Dom validateConfiguration(Plugin plugin, String goal, Xpp3Dom config, Log log) {
         Xpp3Dom goalConfig;
         String executionGoal = plugin.getArtifactId() + ":" + goal;
+        System.out.println("ExecutionGoal: " + executionGoal);
         switch (executionGoal) {
         case "liberty-maven-plugin:create":
             config = convertLibertyAlias(config);
@@ -292,6 +314,12 @@ public class ExecuteMojoUtil {
             break;
         case "maven-ear-plugin:generate-application-xml":
             goalConfig = stripConfigElements(config, EAR_GENERATE_APPLICATION_XML_PARAMS);
+            break;
+        case "maven-ear-plugin:ear":
+            goalConfig = stripConfigElements(config, EAR_PARAMS);
+            break;
+        case "maven-war-plugin:war":
+            goalConfig = stripConfigElements(config, WAR_PARAMS);
             break;
         default:
             goalConfig = config;

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/ExecuteMojoUtil.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/ExecuteMojoUtil.java
@@ -148,17 +148,6 @@ public class ExecuteMojoUtil {
                     "packagingExcludes", "packagingIncludes", "skinnyModules", "skinnyWars", "skipClassPathModification", "unpackTypes",
                     "useBaseVersion", "useJvmChmod", "version"));
             
-    // https://maven.apache.org/plugins/maven-war-plugin/war-mojo.html
-    private static final ArrayList<String> WAR_PARAMS = new ArrayList<>(
-            Arrays.asList("outputDirectory", "warSourceDirectory", "webappDirectory", "workDirectory", "archive",
-                    "archiveClasses", "attachClasses", "classesClassifier", "classifier", "containerConfigXML", "delimiters",
-                    "dependentWarExcludes", "dependentWarIncludes", "escapeString", "escapedBackslashesInFilePath",
-                    "failOnMissingWebXml", "filteringDeploymentDescriptors", "filters", "includeEmptyDirectories",
-                    "nonFilteredFileExtensions", "outdatedCheckPath", "outputFileNameMapping", "outputTimestamp", "overlays", "packagingExcludes",
-                    "packagingIncludes", "primaryArtifact", "recompressZippedFiles", "resourceEncoding", "skip", "supportMultiLineFiltering",
-                    "useDefaultDelimiters", "useJvmChmod", "warSourceExcludes", "warSourceIncludes", "webResources", "webXml"));
-
-
     // https://maven.apache.org/plugins/maven-ear-plugin/generate-application-xml-mojo.html
     private static final ArrayList<String> EAR_GENERATE_APPLICATION_XML_PARAMS = new ArrayList<>(
             Arrays.asList("outputFileNameMapping", "tempFolder", "workDirectory", "applicationId", "applicationName",
@@ -271,7 +260,6 @@ public class ExecuteMojoUtil {
     private static Xpp3Dom validateConfiguration(Plugin plugin, String goal, Xpp3Dom config, Log log) {
         Xpp3Dom goalConfig;
         String executionGoal = plugin.getArtifactId() + ":" + goal;
-        System.out.println("ExecutionGoal: " + executionGoal);
         switch (executionGoal) {
         case "liberty-maven-plugin:create":
             config = convertLibertyAlias(config);
@@ -317,9 +305,6 @@ public class ExecuteMojoUtil {
             break;
         case "maven-ear-plugin:ear":
             goalConfig = stripConfigElements(config, EAR_PARAMS);
-            break;
-        case "maven-war-plugin:war":
-            goalConfig = stripConfigElements(config, WAR_PARAMS);
             break;
         default:
             goalConfig = config;

